definitions.h

```diff
@@ -214,37 +221,234 @@ class BufferImage : public Buffer2D<PIXEL>
+}
+
+// Multiplication of 4x4 by 4x1 matrix
+Vertex operator* (const Matrix& lhs, const Vertex& rhs) {
+    Vertex result = {
+          lhs[0] * rhs.x
+        + lhs[1] * rhs.y
+        + lhs[2] * rhs.z
+        + lhs[3] * rhs.w,
+
+          lhs[4] * rhs.x
+        + lhs[5] * rhs.y
+        + lhs[6] * rhs.z
+        + lhs[7] * rhs.w,
+
+          lhs[8] * rhs.x
+        + lhs[9] * rhs.y
+        + lhs[10] * rhs.z
+        + lhs[11] * rhs.w,
+
+          lhs[12] * rhs.x
+        + lhs[13] * rhs.y
+        + lhs[14] * rhs.z
+        + lhs[15] * rhs.w,
+    };
+    return result;
+}
+
+// Multiplication of 4x4 by 4x4 matrix
+Matrix operator* (const Matrix& lhs, const Matrix& rhs) {
+    Matrix result;
+    
+    // Loop over every cell in result
+    for (int i = 0; i < 16; i++) {
+        int r = 4 * (i / 4);
+        int c = i % 4;
+        double sum = 0;
+
+        // Loop 4 times (row is 4 long, col is 4 long)
+        for (int j = 0; j < 4; j++) {
+            sum += (lhs[r] * rhs[c]);
+            r++;
+            c += 4;
+        }
+        result[i] = sum;
+    }
+
+    return result;
+}
+
+/***************************************************
+ * Matrix transformation functions - return a
+ * matrix that has the specified transformation
+ **************************************************/
+void translateMatrix(Attributes& attr, Vertex v) {
+    attr.matrix[3]  = v.x;
+    attr.matrix[7]  = v.y;
+    attr.matrix[11] = v.z;
+}
+
+void scaleMatrix(Attributes& attr, Vertex v) {
+    attr.matrix[0]  = v.x;
+    attr.matrix[5]  = v.y;
+    attr.matrix[10] = v.z;
+}
+
+// This function takes an angle in degrees and converts it to radians
+void rotateZMatrix(Attributes& attr, double angle) {
+    angle = angle * M_PI / 180.0;
+    // compute trig functions here so we dont have to do them twice
+    double sinangle = sin(angle);
+    double cosangle = cos(angle);
+
+    attr.matrix[0] = cosangle;
+    attr.matrix[1] = -sinangle;
+    attr.matrix[4] = sinangle;
+    attr.matrix[5] = cosangle;
+}
+
+// These functions are not used in the week05 project, but they are still here
+void rotateXMatrix(Attributes& attr, double angle) {
+    angle = angle * M_PI / 180.0;
+    // compute trig functions here so we dont have to do them twice
+    double sinangle = sin(angle);
+    double cosangle = cos(angle);
+
+    attr.matrix[5]  = cosangle;
+    attr.matrix[6]  = -sinangle;
+    attr.matrix[9]  = sinangle;
+    attr.matrix[10] = cosangle;
+}
+
+void rotateYMatrix(Attributes& attr, double angle) {
+    angle = angle * M_PI / 180.0;
+    // compute trig functions here so we dont have to do them twice
+    double sinangle = sin(angle);
+    double cosangle = cos(angle);
+
+    attr.matrix[0]  = cosangle;
+    attr.matrix[2]  = sinangle;
+    attr.matrix[8]  = -sinangle;
+    attr.matrix[10] = cosangle;
+}
+
+// Does all three previous operations at once in right to left order
+void STRMatrix(Attributes& attr, Vertex s, Vertex t, double r) {
```

Comment:
4x4 matrix multiplication is looking good. This looks like a great way to perform all of the transformations without making things too complicated. Good job.